WIP: enable diff mode for AWS modules

s-hertel opened the pull request:
##### SUMMARY
WIP. Unit tests, integration tests, utilization to come.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/aws/core.py

##### ANSIBLE VERSION
```
2.6
```

s-hertel committed:
WIP: enable diff mode for AWS modules

s-hertel committed:
Make python 2.6 compatible

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
+            return [self.create_response_dict(i) for i in item]
+        # from collections import OrderedDict fails on Python 2.6
+        # botocore uses its own OrderedDict class that is compatible with Python 2.6
+        elif item.__class__.__name__ == 'OrderedDict':
```
Why not check the actual type? In botocore we'll use the backport from the `ordereddict` package on 2.6. So `botcore.compat.OrderedDict` will be the version from `collections` on versions that support it. Are there other versions that you're looking to support?

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
             self._module.fail_json(msg=message, exception=last_traceback,
                                    **camel_dict_to_snake_dict(response))
+
+    def create_response_dict(self, item):
```
It looks like all this method is doing is un-ordering dicts, so something like `convert_response` may be a better name.

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
+        elif isinstance(response, list):
+            return [self.populate_response(i, extra_output) for i in response]
+        elif isinstance(response, dict):
+            for output_key, v in dict(response).items():
```
You shouldn't need to cast this to a dict since you've already asserted it is one.

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
+                    # Careful. If doing diffs with deeply nested structures with repeat keys this will
+                    # do weird things. Safer to pass extra_output as a complex structure in cases like that.
+                    # TODO: make extra_output smarter
+                    response[output_key] = self.populate_response(v, extra_output)
```
What's the expected structure of `extra_output`? It looks like you're not traversing it.

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
+        # Generate response dict
+        resp = self.create_response_dict(output_args)
+        if extra_output:
+            resp = self.populate_response(dict(resp), extra_output)
```
This will fail if the response type isn't a dict. If it is a dict, you don't need to cast it since you've already done that in `create_response_dict`.

Also both of these methods traverse the dictionaries. That's probably not an issue, but it might makes sense to consolidate them so you're only traversing it once.

JordonPhillips commented on lib/ansible/module_utils/aws/core.py:
```diff
@@ -159,3 +167,120 @@ def fail_json_aws(self, exception, msg=None):
+        if method:
+            if not any(method.startswith(r) for r in ('get', 'describe', 'list')):
+                self.warn("This may be an implementation error of diff mode. The compare "
+                          "method should be a read-only method. Setting 'before' to an empty dict.")
```
Yeah, I wish we had better metadata around this. You can look at the http method, but POST might be used for read-only stuff for some services. I'm *pretty sure* GET methods don't modify state, but not 100%.

s-hertel committed:
rename method

remove redundant casting to dict

Use 2.6 compatible OrderedDict

add self._diff and self.difflist to AnsibleAWSModule

s-hertel committed:
Add diff mode to ec2_group
